### ISSUE TITLE
Actor is permanently blocked when registering reminders with MongoDB

### ISSUE THREAD

**Issue by MoeweX** [dapr/components-contrib#1355]
## Expected Behavior

When I ask an actor to register multiple reminders, I want him to register all these reminders.

## Actual Behavior

The actors stops doing anything after the first reminder registration. This also blocks the application that invoked the actor's method for reminder registration.  After a couple of minutes, an exception appears that also kills the application, see output: The first line is from the registerReminders method below:

```bash
== APP == Created reminder 0
== APP == 2021-11-17 16:56:37,537 {HH:mm:ss.SSS} [http-nio-3001-exec-4] ERROR o.a.c.c.C.[.[.[.[dispatcherServlet] - Servlet.service() for servlet [dispatcherServlet] threw exception
== APP == io.grpc.StatusRuntimeException: UNKNOWN: could not parse actor reminders: json: cannot unmarshal array into Go value of type actors.Reminder
== APP == 	at io.grpc.Status.asRuntimeException(Status.java:535)
== APP == 	at io.grpc.stub.ClientCalls$UnaryStreamToFuture.onClose(ClientCalls.java:533)
== APP == 	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:553)
== APP == 	at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:68)
== APP == 	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:739)
== APP == 	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:718)
== APP == 	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
== APP == 	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
== APP == 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
== APP == 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
== APP == 	at java.base/java.lang.Thread.run(Thread.java:829)
== APP == 2021-11-17 16:56:37,537 {HH:mm:ss.SSS} [http-nio-3001-exec-4] ERROR o.a.c.c.C.[.[.[.[dispatcherServlet] - Servlet.service() for servlet [dispatcherServlet] in context with path [] threw exception [Request processing failed; nested exception is java.lang.RuntimeException: java.lang.reflect.InvocationTargetException] with root cause
== APP == io.grpc.StatusRuntimeException: UNKNOWN: could not parse actor reminders: json: cannot unmarshal array into Go value of type actors.Reminder
== APP == 	at io.grpc.Status.asRuntimeException(Status.java:535)
== APP == 	at io.grpc.stub.ClientCalls$UnaryStreamToFuture.onClose(ClientCalls.java:533)
== APP == 	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:553)
== APP == 	at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:68)
== APP == 	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:739)
== APP == 	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:718)
== APP == 	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
== APP == 	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
== APP == 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
== APP == 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
== APP == 	at java.base/java.lang.Thread.run(Thread.java:829)
```

## Steps to Reproduce the Problem

Tested with Dapr 1.5.0., I ran the same workload with Redis locally and CosmosDB; both worked fine. Only for MongoDB I have this issue.

```java
// This method is invoked from the outside after the actor was created the first time.
@Override
public double registerReminders(int numberOfReminders) {
  Duration dueTime = Duration.ofDays(1);
  Duration period = Duration.ofDays(1);
  List<Long> registrationTimes = new ArrayList<>(numberOfReminders);

  for (int i = 0; i < numberOfReminders; i++) {
    long start = System.nanoTime();
    super.registerReminder("benchmark-" + i, getId() + "-" + i, dueTime, period).block();
    registrationTimes.add(System.nanoTime() - start);
    System.out.println("Created reminder " + i);
  }

  return registrationTimes.stream().mapToDouble(l -> l).average().orElseThrow();
}
```

I use Mongo Atlas, my config looks like so:

```yaml
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: statestore
spec:
  type: state.mongodb
  version: v1
  metadata:
    - name: server
      value: REMOVED
    - name: databaseName
      value: daprStore
    - name: collectionName
      value: daprCollection
    - name: operationTimeout
      value: 10s
    - name: actorStateStore
      value: "true"
```

The first reminder can be found in the database though:
```json
{
  "actorid": "Actor-0",
  "actortype": "BenchmarkActor",
  "name": "benchmark-0",
  "data": {
    "$binary": {
      "base64": "IkFjdG9yLTAtMCI=",
      "subType": "00"
    }
  },
  "period": "24h0m0s0ms",
  "duetime": "24h0m0s0ms",
  "registeredtime": "2021-11-18T16:41:57+01:00",
  "expirationtime": ""
}
```
Interestingly it looks different to what is found in redis:
```json
{
  "actorID": "Actor-0",
  "actorType": "BenchmarkActor",
  "name": "benchmark-0",
  "data": "IkFjdG9yLTAtMCI=",
  "period": "24h0m0s0ms",
  "dueTime": "24h0m0s0ms",
  "registeredTime": "2021-11-18T17:05:03+01:00"
  }
```

## Release Note

RELEASE NOTE:

**Comment by ravi-k4** [comment]
I am seeing the same error, any update about this issue?

**Comment by yaron2** [comment]
@MoeweX apologies this has taken so long to to get to.

@ravi-k4 @MoeweX can you please try this with Dapr 1.4 and confirm if you're still encountering the issue? I'm trying to understand if this was a regression in this commit: https://github.com/dapr/components-contrib/commit/e369fec17e25d9b128d20ab11356727ed3f97081.

**Comment by mukundansundar** [comment]
I can reproduce this issue. 

The issue is the before the change that Yaron pinged above, we were storing Reminders data inside MongoDB as string.
See log below,

```
 db.actorCollection.find({})
{ "_id" : "actors||DemoActor", "_etag" : "85dc089b-7fbf-4148-b376-bf364f174825", "value" : "[{\"actorID\":\"demoActor1000\",\"actorType\":\"DemoActor\",\"name\":\"testReminder1\",\"data\":1002,\"period\":\"1m\",\"dueTime\":\"1m\",\"registeredTime\":\"2022-03-23T12:19:02+05:30\"}]" }
{ "_id" : "actors||TestActor", "_etag" : "7a171bf8-e864-44a9-84ed-6bb2801af346", "value" : "[{\"actorID\":\"testActor1000\",\"actorType\":\"TestActor\",\"name\":\"testReminder100\",\"data\":{\"key1\":\"val1\",\"key2\":\"val2\"},\"period\":\"30s\",\"dueTime\":\"30s\",\"registeredTime\":\"2022-03-23T12:19:12+05:30\"}]" }
{ "_id" : "DemoActor||demoActor1000||testReminder1", "_etag" : "c3fb0ec3-37a3-452d-bb11-fefff2d4015f", "value" : "{\"lastFiredTime\":\"2022-03-23T12:20:02+05:30\",\"repetitionLeft\":-1}" }
{ "_id" : "TestActor||testActor1000||testReminder100", "_etag" : "f5f66657-d186-4fa5-9c75-6210c67bf2c5", "value" : "{\"lastFiredTime\":\"2022-03-23T12:20:12+05:30\",\"repetitionLeft\":-1}" }
> db.actorCollection.aggregate( 
...     [ 
...         { "$project": { "fieldType": {  "$type": "$value"  } } } 
...     ]
... )
{ "_id" : "actors||DemoActor", "fieldType" : "string" }
{ "_id" : "actors||TestActor", "fieldType" : "string" }
{ "_id" : "DemoActor||demoActor1000||testReminder1", "fieldType" : "string" }
{ "_id" : "TestActor||testActor1000||testReminder100", "fieldType" : "string" }
```

With 1.5.1, this changes 

```
{ "_id" : "actors||TestActor", "_etag" : "05f50a99-ada6-424a-83f9-7c8c41315c23", "value" : [ { "actorid" : "testActor1000", "actortype" : "TestActor", "name" : "base64Data1", "data" : "eyJwcm9wZXJ0eUEiOiJpbmxpbmUiLCJwcm9wZXJ0eUIiOjEwMDB9", "period" : "0h0m30s0ms", "duetime" : "0h0m30s0ms", "registeredtime" : "2022-03-23T15:44:04+05:30", "expirationtime" : "" } ] }
{ "_id" : "TestActor||testActor1000||base64Data1", "_etag" : "1aa8404a-7b64-444a-9545-69dff7cdfc39", "value" : { "lastfiredtime" : "2022-03-23T15:49:34+05:30", "repetitionleft" : -1 } }

{ "_id" : "actors||TestActor", "fieldType" : "array" }
{ "_id" : "TestActor||testActor1000||base64Data1", "fieldType" : "object" }
```

The error log that I saw from a Java test app is as follows : 

```
ERRO[0014] error getting reminders for actor type TestActor: could not parse actor reminders: json: cannot unmarshal array into Go value of type actors.Reminder  app_id=demoactorservice instance=test.local scope=dapr.runtime.actor type=log ver=edge
INFO[0014] placement tables updated, version: 29         app_id=demoactorservice instance=test.local scope=dapr.runtime.actor.internal.placement type=log ver=edge
```

when it was changed back to 1.4.4 after a new reminder was added in 1.5.1, then also the app did not start up. 

```log
ERRO[0011] error getting reminders for actor type TestActor: error decoding key value: cannot decode array into a string type  app_id=demoactorservice instance=test.local scope=dapr.runtime.actor type=log ver=1.4.4
```

**Comment by mukundansundar** [comment]
cc @yaron2 @artursouza 
@dapr/maintainers-components-contrib

**Comment by berndverst** [comment]
On Dapr 1.6 creating a single actor reminder, then restarting the sidecar makes the sidecar init fail.

Data in Mongo looks like this:

```json
{
"_id" : "actors||testactorfeatures",
"_etag" : "e5d3e645-5485-45c9-a789-c8944c20c369",
"value" : [
{
"actorid" : "10",
"actortype" : "testactorfeatures",
"name" : "myReminder",
"data" : "reminderdata",
"period" : "1s",
"duetime" : "1s",
"registeredtime" : "2022-03-23T18:23:20-07:00",
"expirationtime" : ""
}
]
}
```

Error like this:

```
ERRO[0002] error getting reminders for actor type testactorfeatures: could not parse actor reminders: json: cannot unmarshal array into Go value of type actors.Reminder app_id=Shoulderwax-Seer instance=Bernds-MacBook-Pro.local scope=dapr.runtime.actor type=log ver=1.6.0
```

Prior to v1.5.0 data was stored in this format:

```
{
"_id" : "actors||testactorfeatures",
"_etag" : "2ba16d70-2f10-486d-a8ee-bd02bc86e8b1",
"value" : "[{\"actorID\":\"10\",\"actorType\":\"testactorfeatures\",\"name\":\"myReminder\",\"data\":\"reminderdata\",\"period\":\"1s\",\"dueTime\":\"1s\",\"registeredTime\":\"2022-03-23T18:13:55-07:00\"},{\"actorID\":\"20\",\"actorType\":\"testactorfeatures\",\"name\":\"myReminder\",\"data\":\"reminderdata\",\"period\":\"1s\",\"dueTime\":\"1s\",\"registeredTime\":\"2022-03-23T18:14:55-07:00\"},{\"actorID\":\"30\",\"actorType\":\"testactorfeatures\",\"name\":\"myReminder\",\"data\":\"reminderdata\",\"period\":\"1s\",\"dueTime\":\"1s\",\"registeredTime\":\"2022-03-23T18:14:59-07:00\"}]"
}
```

But now it looks like this:

```
{
"_id" : "actors||testactorfeatures",
"_etag" : "86989faa-fb1c-4de3-9132-a29e46a67626",
"value" : [
{
"actorid" : "10",
"actortype" : "testactorfeatures",
"name" : "myReminder",
"data" : "reminderdata",
"period" : "1s",
"duetime" : "1s",
"registeredtime" : "2022-03-23T18:13:55-07:00",
"expirationtime" : ""
},
{
"actorid" : "20",
"actortype" : "testactorfeatures",
"name" : "myReminder",
"data" : "reminderdata",
"period" : "1s",
"duetime" : "1s",
"registeredtime" : "2022-03-23T18:14:55-07:00",
"expirationtime" : ""
},
{
"actorid" : "30",
"actortype" : "testactorfeatures",
"name" : "myReminder",
"data" : "reminderdata",
"period" : "1s",
"duetime" : "1s",
"registeredtime" : "2022-03-23T18:14:59-07:00",
"expirationtime" : ""
},
{
"actorid" : "400",
"actortype" : "testactorfeatures",
"name" : "myReminder",
"data" : "reminderdata",
"period" : "1s",
"duetime" : "1s",
"registeredtime" : "2022-03-23T18:18:02-07:00",
"expirationtime" : ""
}
]
}
```

We somehow should try taking the BSON / relaxed JSON and wrapping that in a string. That might resolve the issue.

**Comment by mukundansundar** [comment]
> We somehow should try taking the BSON / relaxed JSON and wrapping that in a string. That might resolve the issue.

Tried taking the JSON value and wrapping that as string ... that did not work ...

Did some more investigation and found the following: 

The reminders list is converted from string(stored in 1.4.4) to array(1.5+) if a new reminder is written. 
Have to check for deletions but I assume they will also be the same. 
```
# Type details from mongo
{ "_id" : "actors||TestActor", "fieldType" : "array" }
```

For the reminder last fire information, by default even if no new reminder is created, once 1.5 + version of 
Dapr is used and connects to mongo, the type automatically is overwritten from string to object. 
```
# Type details from mongo
{ "_id" : "TestActor||testActor1000||base64Data1", "fieldType" : "object" }
```

As seen from above the type of the reminder object stored in MongoDB in 1.5 + is type array or primitive.A (golang mongo Driver) and not primitive.D (object type). 

Output of that value as a string is as follows
``` 
[[{actorid testActor1000} {actortype TestActor} {name base64Data1} 
{data eyJwcm9wZXJ0eUEiOiJpbmxpbmUiLCJwcm9wZXJ0eUIiOjEwMDB9} {period 0h0m30s0ms}
 {duetime 0h0m30s0ms} {registeredtime 2022-03-24T20:39:12+05:30} {expirationtime }] 
 [{actorid testActor1000} {actortype TestActor} {name base64Data2} {data eyJwcm9wZXJ0eUEiOiJpbmxpbmUiLCJwcm9wZXJ0eUIiOjEwMDB9} 
 {period 0h0m2s0ms} {duetime 0h0m5s0ms} {registeredtime 2022-03-24T20:46:41+05:30} {expirationtime }]]
```

 We need to find a way to convert `primitive.A` type data into string version of the data with escaped values. 
 Example: 
```
 { "_id" : "actors||DemoActor", "_etag" : "85dc089b-7fbf-4148-b376-bf364f174825", 
 "value" : "[{\"actorID\":\"demoActor1000\",\"actorType\":\"DemoActor\",\"name\":\"testReminder1\",\"data\":1002,\"period\":\"1m\",
 \"dueTime\":\"1m\",\"registeredTime\":\"2022-03-23T12:19:02+05:30\"}]" }
```

**Comment by berndverst** [comment]
This has been fixed in #1616 we will also need to cherry pick that into Dapr v1.6 and maybe v1.5.

The Dapr v1.7 will address this.

**Comment by berndverst** [comment]
Hotfixes 1.5.2 and 1.6.1 are now released and address this. Please upgrade to these versions or use the upcoming 1.7.0 release.
